### ISSUE TITLE
Web app issues

### ISSUE THREAD

**Issue by DRSDavidSoft** [peachpiecompiler/peachpie#859]
I've run into a couple of issues while trying to build an existing Web project with Peach Pie, and I'd appreciate it if I could get some help on this.

1. [x] **#861** Is `DEBUG` a special or reserved constant?  This works fine in PHP:
    ```php
    <?PHP
    echo "start";
    define('DEBUG', 1);
    echo "end";

    // curl localhost:5004 → startend
    ```

    However, in Peach Pie, it stops and exits right after the `define()` statement:
    `start`

2. [x] [[SOLVED]](https://github.com/peachpiecompiler/peachpie/issues/859#issuecomment-707299244) The `pdo_mysql` does not seem to be present within Peach Pie.  Is this normal/intentional? My project fails to start if it's not detected:

    ```php
    extension_loaded('pdo_mysql'); // false on PeachPie
    ```

3. [x] **#860** How can I replicate the following IIS `web.config` configuration for URL rewrite?
    ```xml
            <rule name="RewriteRule" enabled="true" stopProcessing="true">
                <match url="^(.*)$" ignoreCase="true" />
                <conditions>
                    <add input="{REQUEST_FILENAME}" matchType="IsFile" ignoreCase="false" negate="true" />
                    <add input="{REQUEST_FILENAME}" matchType="IsDirectory" ignoreCase="false" negate="true" />
                </conditions>
                <action type="Rewrite" url="index.php/{R:1}" appendQueryString="true" />
            </rule>
    ```

    This does not seem to work:
    ```cs
            var options = new RewriteOptions()
                .AddRewrite(@"^(.*)$", "index.php/$1", skipRemainingRules: true);

            app.UseRewriter(options);

    ```

4. [x] **#863** `json_encode()` on `simplexml_load_string()` objects is not working

There are more issues, but any help would be appreciated on these

**Comment by jakubmisek** [comment]
1. you are right, seems `DEBUG` is already defined automatically when building the project in debug configuration. So defining it in users' code is unnecessary. For the clarification, it does not "stop and exit right after the define()", I'm sure it throws an exception.

2. add a package reference to `Peachpie.Library.PDO.MySQL` in your `.msbuildproj` project file:
  ```xml
  <PackageReference Include="Peachpie.Library.PDO.MySQL" Version="$(PeachpieVersion)" />
  ```
  sorry it has not been documented yet!

3. the rewrite rule seems correct to me (as described at https://docs.microsoft.com/en-us/aspnet/core/fundamentals/url-rewriting)

**Comment by DRSDavidSoft** [comment]
1. Thanks, good to know!  

    Is there any way to catch the exception? This doesn't work:  

    ```php
    <?php
    
    echo "start\n";
    try { define('DEBUG', 1); }
    catch ( Exception $e ) { echo $e->getMessage(); }
    echo "end\n";

   `curl localhost:5004 -v` → `start * Connection #0 to host localhost left intact`

2. No worries, thanks again for the info!

3. I'm not sure why, but Kestrel throws a 404 when doing a request like this:
    ```batch
    $ curl -v localhost:5004/foo_bar
    
    < HTTP/1.1 404 Not Found
    ...
    ```

**Comment by jakubmisek** [comment]
1. seems it does not behave the same as in PHP; currently, it throws \System\InvalidOperationException
  ```php
  catch (  \System\InvalidOperationException $e ) { echo $e->Message; }
  ```
  but better to just handle the exception in Visual Studio or VS Code when debugging.
  We should probably create a new issue so `define()` would result in the same error as in PHP.

3. I noticed, you rewrite the URL into `"index.php/$1"` which is then not handled since "index.php/" is a directory here. I'm wondering, how is this handled on Apache? Does it end up in `index.php` script file?

**Comment by DRSDavidSoft** [comment]
1. Thanks, I'll look into the documentation for this https://docs.peachpie.io/scenarios/intermediate/debugging-peachpie/

3. Oh, usually you can access a path like `/project/index.php/path/to/some/url` in PHP within `$_SERVER['PATH_INFO']`:

```php
print_r($_SERVER['PATH_INFO']);
```

```bash
$ curl localhost/projecta/index.php/path/to/some/url

# returns: /path/to/some/url
```

The PHP [documentation](https://www.php.net/manual/en/reserved.variables.server.php) says;
> 'PATH_INFO'
> Contains any client-provided pathname information trailing the actual script filename but preceding the query string, if available. For instance, if the current script was accessed via the URL http://www.example.com/php/path_info.php/some/stuff?foo=bar, then $_SERVER['PATH_INFO'] would contain /some/stuff.

This SO [post](https://stackoverflow.com/a/2261971/1454514) suggests it "is an environment variable set by Apache", but it actually is "part of of the definition for CGI" [RFC3875#section-4.1.5](https://tools.ietf.org/html/rfc3875#section-4.1.5), and it's supported both by IIS+PHP, and the internal PHP server (`php -S`).

I use the URL rewrite option in both Apache and IIS to remove the `index.php` file, so `/path/index.php/foo/bar` becomes `/path/foo/bar`.  Here's the `.htaccess` directive that matches the `web.config` that I posted earlier:

```
	RewriteEngine On
	RewriteCond %{REQUEST_FILENAME} !-f  
	RewriteCond %{REQUEST_FILENAME} !-d  
	RewriteRule ^(.*) index.php/$1 [L,QSA]
```

The `IsFile` (`-f` in Apache) or `IsDirectory` (`-d` in Apache) do not seem to interfere with the `PATH_INFO` method, although I think this might be missing from Peach Pie + Kestrel.  I do not see why it might return a 404, though.

**Comment by jakubmisek** [comment]
1. you don't have to debug PeachPie itself, just your application. Simply open your PHP project in VSCode or Visual Studio.

2. I see! PeachPie's asp.net core integration does not handle that now, and `PATH_INFO` is populated with an incorrect value I suppose. (created an issue at https://github.com/peachpiecompiler/peachpie/issues/860)

I don't see the rest of your C# code, just to be sure, do you have following there?
```c#
app.UsePhp();
```

**Comment by DRSDavidSoft** [comment]
I've actually cloned the [`web-application`](https://github.com/iolevel/peachpie-samples/tree/master/web-application) directory from `peachpie-samples`:

This section is unmodified:
```cs
app.UsePhp("/", rootPath: contentPath);
```

![Uploading VS-Code-PeachPie.png…]()

**Comment by jakubmisek** [comment]
thanks! looks good.

anyways; the AddRewrite( ... , "index.php/$1" ) broke the app, since now everything rewrites to "index.php/something" which is never handled. (that's why localhost:5004/index.php returns 404)

**Comment by DRSDavidSoft** [comment]
@jakubmisek I get the same `HTTP/1.1 404 Not Found` on `localhost:5004/index.php -v` **even if** the rewrite portion is commented out:
```cs
            // var options = new RewriteOptions()
            //     .AddRewrite(@"^(.*)$", "index.php/$1", skipRemainingRules: true);

            // app.UseRewriter(options);
```

**Comment by jakubmisek** [comment]
@DRSDavidSoft but `localhost:5004` gives you the result as expected?

**Comment by DRSDavidSoft** [comment]
@jakubmisek yes -- only if the rewrite section is commented out.

**Comment by jakubmisek** [comment]
I see, `localhost:5004/index.php` does not work indeed

**Comment by DRSDavidSoft** [comment]
@jakubmisek There is another issue that I'll post in #863

**Comment by DRSDavidSoft** [comment]
Subscribing to the new (sub)-issues for when they are resolved.

**Comment by jakubmisek** [comment]
update; your URL rewriting rule is correct, with https://github.com/peachpiecompiler/peachpie/issues/860 fixed, it will work as expected.

You can try the latest build if you add NuGet source feed: https://ci.appveyor.com/nuget/peachpie
and change peachpie version in your project files from "1.0.0-preview2" to "1.0.0-appv4183" (or newer)

**Comment by DRSDavidSoft** [comment]
@jakubmisek Just tried `1.0.0-appv4183`, can confirm it works; and whoa, it's fast! 😃  Thank you for your efforts.  I'm finding more issues that need to be fixed, so I'm looking forward to see the current ones being closed. 🍻

**Comment by DRSDavidSoft** [comment]
@jakubmisek There is a tiny bit of inconsistency, though.  The `PATH_INFO` should be `""` (empty string) when issuing a call like this: `$ curl localhost:5004`.  Currently, PeachPie sets it to `"/"` when such a request is made.

You can test this out with both of the IIS and Apache config files I posted above, (`web.config` and `.htaccess` respectively.)

This doesn't matter to me, since I `trim($_SERVER['PATH_INFO'])` in my codebase now that I found this out, but I thought I'd mention this if you'd like the behavior to _exactly_ match PHP.  Thanks!

**Comment by DRSDavidSoft** [comment]
@jakubmisek I noticed there is a second issue, where if

- The `RewriteOptions()` is commented out
- _And_ `index.php` exist in the root of `website/`

then it seems that the requests are automatically rewritten to `index.php`, no matter what the request is!

**Comment by jakubmisek** [comment]
@DRSDavidSoft seems the builtin php server behaves a bit differently as well.

**Comment by jakubmisek** [comment]
> @jakubmisek There is a tiny bit of inconsistency, though. The `PATH_INFO` should be `""` (empty string) when issuing a call like this: `$ curl localhost:5004`. Currently, PeachPie sets it to `"/"` when such a request is made.

PeachPie leaves `$_SERVER['PATH_INFO']` unset in this case, just tested. But we'll will try again

**Comment by DRSDavidSoft** [comment]
@jakubmisek I can confirm that the `$_SERVER['PATH_INFO']` value is correct now.

There is a final issue with URL Rewrite, where a request to an _existing_ static file will be rewritten to `index.php`, where it shouldn't.  I'm not sure if this is something expected in PeachPie, and the Microsoft documentation lacks any info on exclusion of static files, as it seems that it should be done by default.  I asked for more information on SO [here](https://stackoverflow.com/q/64342986/1454514).

```cs
// with URL rewrite enabled:
info: Microsoft.AspNetCore.Rewrite.RewriteMiddleware[10]
      Request was rewritten to index.php/view/assets/fonts/webfonts.css

// with URL rewrite disabled:
info: Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware[2]
      Sending file. Request path: '/view/assets/fonts/webfonts.css'. Physical path: 'C:\Users\David\AppData\Local\Temp\peachpie-samples\web-application\website\view\assets\fonts\webfonts.css'
```

**Comment by jakubmisek** [comment]
I suppose that's how `RewriteOptions` and `Microsoft.AspNetCore.Rewrite` works. `AddRewrite` simply changes the request path according to your reqexp. There are several ways how to make it work:

- move the `app.UseStaticFiles();` above the `app.UseRewriter(options);`
- add your own Rule to `RewriteOptions` that would check you are requesting existing static file and would skip the remaining rules (I'm wondering if there is something like this already implemented in aspnetcore)

https://stackoverflow.com/questions/42869686/asp-core-url-rewrite-not-working-when-using-addiisurlrewrite
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/url-rewriting
 I couldn't find more information on this so far

**Comment by DRSDavidSoft** [comment]
@jakubmisek Thank you for going throw the documentation, what I gathered from this is that the order of "using" the middleware matter.  However, I still have some issues/inconsistencies.

Can I submit some tests for ASP.NET Core + PeachPie to check for consistency with PHP?  If possible, please add some stubs in a new branch where I could make a PR.

I'm closing this issue now, and will post the remaining problems in a new issue. Thanks!

**Comment by jakubmisek** [comment]
@DRSDavidSoft thank you, please post new issues.

The URL rewriting is however out of the scope of peachpie itself, we can just discuss possible approaches and best practices.
